Jun's TUV-x cloud optical property calculations

mattldawson opened the pull request:
@jzhan166 - could you add a description here of how you're calculating the cloud optical properties? I'll use that in the description for the CAM PR.

fvitt committed:
use tuvx lib branch; hack for clear sky; turn on jno; add history fields

	modified:   Externals_CAM.cfg
	modified:   src/chemistry/mozart/mo_photo.F90
	modified:   src/chemistry/mozart/mo_tuvx.F90

	new file:   src/chemistry/pp_waccm_ma/tuvx_config.json
	new file:   src/chemistry/pp_waccm_ma/tuvx_config.json_121.65
	new file:   src/chemistry/pp_waccm_tsmlt_mam5/tuvx_config.json
	new file:   src/chemistry/pp_waccm_tsmlt_mam5/tuvx_config.json_121.65

fvitt committed:
Jun's updates

	modified:   Externals_CAM.cfg
	modified:   src/chemistry/pp_waccm_tsmlt_mam5/tuvx_config.json

fvitt committed:
implement Chuck's method of interpolating rrtmg aerosol optical properties

	modified:   src/chemistry/mozart/mo_tuvx.F90
	modified:   src/physics/rrtmg/radiation.F90

fvitt committed:
get cloud raditative properties from RRTMG

	modified:   src/chemistry/mozart/mo_tuvx.F90
	modified:   src/physics/rrtmg/radiation.F90

fvitt committed:
fix typo error for cldg

        modified:   src/chemistry/mozart/mo_tuvx.F90

fvitt committed:
some clean up on the logic to turn on effects of clouds and aerosols

	modified:   src/chemistry/mozart/mo_tuvx.F90
	modified:   src/chemistry/pp_waccm_ma/tuvx_config.json
	modified:   src/chemistry/pp_waccm_tsmlt_mam5/tuvx_config.json
	new file:   src/chemistry/pp_waccm_ma_mam5/tuvx_config.json

mattldawson commented on src/chemistry/mozart/mo_photo.F90:
```diff
@@ -823,9 +823,12 @@ subroutine table_photo( photos, pmid, pdel, temper, zmid, zint, &
+          !cld_mult(:) = esfact * cld_mult(:)
+
+          ! Clear Sky
+          cld_mult(:) = 1._r8
```
I would suggest we not modify the existing TUV (lookup table) calculations, so that we don't end up with answer changes in the regression tests. Is it possible to remove these changes?

mattldawson commented on src/chemistry/mozart/mo_tuvx.F90:
```diff
@@ -75,10 +82,10 @@ module mo_tuvx
+   logical :: do_clouds  = .true.  ! indicates whether cloud optical properties
                                    !   should be calculated and used in radiative
                                    !   transfer calculations
 
```
These should default to `.false.` they will be set to true in the code if the needed data is available and the configuration file does not disable them.

mattldawson commented on src/chemistry/mozart/mo_tuvx.F90:
```diff
@@ -89,7 +96,7 @@ module mo_tuvx
 
    ! Information needed to do special NO photolysis rate calculation
-   logical :: do_jno     = .false. ! Indicates whether to calculate jno
+   logical :: do_jno     = .true. ! Indicates whether to calculate jno
```
This should default to `.false.`. It will be set to true if the configuration file specifies to calculate the NO photolysis rate constant.

mattldawson commented on src/chemistry/mozart/mo_tuvx.F90:
```diff
@@ -132,6 +139,16 @@ module mo_tuvx
+  integer :: swcldtauwg_idx = -1
+  type (interp_type) :: interp_wgts
+  real(r8) :: rrtmg_wavelength(nswbands-1)
+  integer :: nwave
```
Please provide descriptions for these and consider using readable variable names.

mattldawson commented on src/chemistry/mozart/mo_tuvx.F90:
```diff
@@ -281,6 +300,10 @@ subroutine tuvx_init( photon_file, electron_file, max_solar_zenith_angle, pbuf2d
+      real(r8) :: wvl_low(nswbands)
+      real(r8) :: wvl_high(nswbands)
+      real(r8), allocatable :: wc(:)
+
```
please provide descriptions for these and consider using readable variable names.

mattldawson commented on src/chemistry/mozart/mo_tuvx.F90:
```diff
@@ -413,6 +436,15 @@ subroutine tuvx_init( photon_file, electron_file, max_solar_zenith_angle, pbuf2d
       end do
+
+      ! get number of TUVx wave bins
+      wavelength => cam_grids%get_grid( "wavelength", "nm" )
```
`wavelength` will need deallocated to avoid a memory leak (it is a copy of the grid, as opposed to a pointer to the actual grid)

mattldawson commented on src/chemistry/mozart/mo_tuvx.F90:
```diff
@@ -590,13 +648,20 @@ subroutine tuvx_get_photo_rates( state, pbuf, ncol, lchnk, height_mid, &
+
+         call get_aerosol_optical_properties2( tuvx, pbuf, state%ncol, &
+              optical_depth, single_scattering_albedo, asymmetry_factor, &
+              optical_depth_cld, single_scattering_albedo_cld, asymmetry_factor_cld )
```
please replace the `get_aerosol_optical_properties()` function rather than creating a modified copy, if it no longer being used.

mattldawson commented on src/chemistry/mozart/mo_tuvx.F90:
```diff
@@ -887,20 +962,39 @@ subroutine initialize_diagnostics( this )
          call addfld( "tuvx_"//diagnostics( i_label )%name_, (/ 'lev' /), 'A', 'sec-1', &
-            'photolysis rate constant' )
+                      trim(diagnostics( i_label )%name_)//' photolysis rate constant' )
+       !  call add_default("tuvx_"//diagnostics( i_label )%name_,3,' ')
```
please remove commented out code

mattldawson commented on src/chemistry/mozart/mo_tuvx.F90:
```diff
@@ -887,20 +962,39 @@ subroutine initialize_diagnostics( this )
+
+      do ipht = 1, phtcnt
+         call addfld('tuvcam_'//trim(rxt_tag_lst(ipht)), (/ 'lev' /), 'A', 'sec-1', 'photolysis rate constant' )
+        ! call add_default('tuvcam_'//trim(rxt_tag_lst(ipht)), 3,' ')
```
please remove commented out code

mattldawson commented on src/chemistry/mozart/mo_tuvx.F90:
```diff
@@ -1727,6 +1809,11 @@ subroutine get_aerosol_optical_properties( this, state, pbuf, optical_depth, &
       ! regrid optical properties to TUV-x wavelength and height grid
       ! =============================================================
       ! TODO is this the correct regridding scheme to use?
+      ! FVITT -- optical props are zero for wavelengths shorter than then RRTMG bands -- < 200 nm ???
```
is this something that needs to be addressed before we submit the PR?

mattldawson commented on src/chemistry/mozart/mo_tuvx.F90:
```diff
@@ -1759,6 +1852,115 @@ subroutine get_aerosol_optical_properties( this, state, pbuf, optical_depth, &
+
+      real(r8), pointer, dimension(:,:,:) :: swcldtau   ! shortwave aerosol tau
+      real(r8), pointer, dimension(:,:,:) :: swcldtauw  ! shortwave aerosol tau * w
+      real(r8), pointer, dimension(:,:,:) :: swcldtauwg ! shortwave aerosol tau * w * g
```
please use consistent descriptions for variables and avoid greek/latin letters as descriptions (i.e. "optical depth" instead of "tau"). Also, consider using readable variable names

mattldawson commented on src/chemistry/mozart/mo_tuvx.F90:
```diff
@@ -1759,6 +1852,115 @@ subroutine get_aerosol_optical_properties( this, state, pbuf, optical_depth, &
+      real(r8) :: gcld(pcols, pver, nwave)
+
+      real(r8) :: swcldw(pcols, pver, nswbands)
+      real(r8) :: swcldg(pcols, pver, nswbands)
```
please provide descriptions with units for all variables, and consider using readable variable names.

mattldawson commented on src/chemistry/pp_waccm_ma/tuvx_config.json_121.65:
```diff
@@ -0,0 +1,1480 @@
+{
```
Is this file needed?

mattldawson commented on src/chemistry/pp_waccm_tsmlt_mam5/tuvx_config.json_121.65:
```diff
@@ -0,0 +1,2006 @@
+{
```
Is this file needed?

fvitt replied:
For `jno_i` we need to revert the `wavelength` value back to 121.65.  The 121.50 value was used to match table-look-up wavelength grid.

So I believe for `jno_i`, we should have:
```
            "name": "jno_i",
            "__reaction": "NO + hv -> NOp + e",
            "cross section": {
               "type": "base",
               "data": {
                  "default value": 0.0,
                  "point values": [
                     { "wavelength": 121.65, "value": 2.0e-18 }
                  ]
               }
            },
```
All our json files need to be checked.

The `tuvx_config.json_121.65` file should be removed after this change.

jzhan166 replied:
Hi Francis, how about for jh2so4
            "name": "jh2so4",
            "__reaction": "H2SO4 + hv -> SO3 + H2O",
            "cross section": {
               "type": "base",
               "data": {
                 "default value": 0.0,
                 "point values": [
                   **{ "wavelength": 121.50, "value":      6.3e-17 },**

fvitt replied:
I missed that. Yes, change it back to 121.65 for jh2so4.

mattldawson commented on src/physics/rrtmg/radiation.F90:
```diff
@@ -145,6 +145,14 @@ module radiation
 integer :: cld_idx      = 0
 integer :: cldfgrau_idx = 0
 
+integer :: swaertau_idx   = -1
+integer :: swaertauw_idx  = -1
+integer :: swaertauwg_idx = -1
+
+integer :: swcldtau_idx   = -1
+integer :: swcldtauw_idx  = -1
+integer :: swcldtauwg_idx = -1
+
```
please provide descriptions for all variables and consider using readable variable names

mattldawson reviewed (commented):
@jzhan166 - looks good! Just a few comments/questions.

jzhan166 committed:
address Matt's comments

Commit:
Merge branch 'develop-add-tuvx' into tuvx_jun_cld

Commit:
update netcdf file names in tuv-x config

fvitt commented:
@mattldawson 
We are not able to build CESM when DEBUG is TRUE.
See this case: 
```
 /glade/derecho/scratch/fvitt/FWHIST_tuvx_jun_cld_test02/bld/atm.bldlog.250918-115934
```

Error message is:
```
ERROR: yaml-cpp library not found at /glade/derecho/scratch/fvitt/FWHIST_tuvx_jun_cld_test02/bld/lib/lib/libyaml-cpp.a
```

fvitt committed:
workaround yaml lib link issue

	modified:   cime_config/buildlib
	modified:   src/chemistry/mozart/mo_photo.F90
	modified:   src/chemistry/pp_waccm_ma/tuvx_config.json
	modified:   src/chemistry/pp_waccm_ma_mam5/tuvx_config.json

fvitt committed:
Initialize aerosol and cloud optical properties to zero

	modified:   src/chemistry/mozart/mo_tuvx.F90

fvitt committed:
add tuvx to carma_mixed_sulfate test; add pbuf fields in mo_tuvx; get opitical props from rrtmgp

	new file:   src/chemistry/pp_waccm_ma_sulfur/tuvx_config.json
	modified:   cime_config/testdefs/testmods_dirs/cam/carma_mixed_sulfate/user_nl_cam
	modified:   src/chemistry/mozart/mo_tuvx.F90
	modified:   src/chemistry/pp_waccm_ma/tuvx_config.json
	modified:   src/chemistry/pp_waccm_ma_mam5/tuvx_config.json
	modified:   src/physics/rrtmg/radiation.F90
	modified:   src/physics/rrtmgp/rrtmgp_inputs_cam.F90

fvitt committed:
add tuvx_config.json for T4S chem

fvitt committed:
implement radiation dose rates diagnostics

	modified:   cime_config/testdefs/testmods_dirs/cam/outfrq9s_waccm_ma_mam4/user_nl_cam
	modified:   src/chemistry/mozart/mo_tuvx.F90
	modified:   src/chemistry/pp_waccm_ma_mam5/tuvx_config.json

fvitt committed:
fix issue with dose rates diags; misc cleanup

	modified:   src/chemistry/mozart/mo_tuvx.F90

fvitt committed:
changes to tuvx lib cmake bld for izumi

	modified:   cime_config/buildlib

fvitt committed:
fix tuvx bld for modeling2

fvitt committed:
more changes to tuvx lib cmake bld for izumi

	modified:   cime_config/buildlib

fvitt committed:
fix izumi NAG builds

	modified:   .gitmodules
	modified:   cime_config/buildlib
	modified:   libraries/tuv-x (new commits)

fvitt committed:
add outfrq9s_tuvx_waccm_ma_mam5 test mod

fvitt committed:
enable clouds and aerosols and turn on dose rates in pp_waccm_tsmlt_mam5/tuvx_config.json
